README.md
# 🌲 Timber exception capturing for Elixir

[![ISC License](https://img.shields.io/badge/license-ISC-ff69b4.svg)](LICENSE.md)
[![Hex.pm](https://img.shields.io/hexpm/v/timber_exceptions.svg?maxAge=18000=plastic)](https://hex.pm/packages/timber_exceptions)
[![Documentation](https://img.shields.io/badge/hexdocs-latest-blue.svg)](https://hexdocs.pm/timber_exceptions/index.html)
[![Build Status](https://travis-ci.org/timberio/timber-elixir-exceptions.svg?branch=master)](https://travis-ci.org/timberio/timber-elixir-exceptions)

The Timber Exceptions library provides enhanced logging of exceptions that occur
in your Elixir software.

## Installation

1. Ensure that you have both `:timber` (version 3.0.0 or later) and `:timber_exceptions` listed
as dependencies in `mix.exs`:

    ```elixir
    def deps do
      [
        {:timber, "~> 3.1"},
        {:timber_exceptions, "~> 2.0"}
      ]
    end
    ```

2. Run `mix deps.get` to install the dependencies.

3. Add the translator in your application's `start/2` function:

    ```elixir
    defmodule MyApp
      use Applictation
      
      def start(_type, _args) do
        # ADD THE FOLLOWING LINE:
        :ok = Logger.add_translator({Timber.Exceptions.Translator, :translate})

        opts = [strategy: :one_for_one, name: MyApp.Supervisor]
        Supervisor.start_link(children, opts)
      end
    end
    ```

## License

This project is licensed under the ISC License - see [LICENSE] for more details.


CHANGELOG.md
# Changelog

This covers changes for versions 2.0 and higher. The changelog for 1.x releases
can be found in the [v1.x
branch](https://github.com/timberio/timber-elixir-exceptions/blob/v1.x/CHANGELOG.md).

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.1.0] - 2019-02-25

### Changed
  - Removed use of Timber-specific custom context and events in favor of simple maps

## 2.0.0 - 2019-01-02

### Added

  - Added a Logger translator to handle exception capturing with Elixir 1.7 and
    OTP 21

### Removed

  - Removed support for versions of Elixir 1.6 and below
  - Removed the Erlang `:error_logger` handler in favor of the Elixir 1.7 Logger
    translator
    
[Unreleased]: https://github.com/timberio/timber-elixir-exceptions/compare/v2.1.0...HEAD
[2.1.0]: https://github.com/timberio/timber-elixir-exceptions/compare/v2.0.0...v2.1.0


config/test.exs
use Mix.Config

config :logger, :handle_otp_reports, false

config :timber,
  api_key: "api_key",
  capture_errors: true


test/support/test_helpers.ex
defmodule Timber.Exceptions.TestHelpers do
  defmacro skip_min_elixir_version(version) do
    quote do
      unless(Version.match?(System.version(), "~> #{unquote(version)}")) do
        @tag :skip
      end
    end
  end

  # This replicates functions that are defined in Timber.TestHelpers but are
  # defined here as well for the purposes of 
  def event_entry_to_log_entry({level, _, {Logger, message, ts, metadata}}) do
    Timber.LogEntry.new(ts, level, message, metadata)
  end

  def event_entry_to_msgpack(entry) do
    log_entry = event_entry_to_log_entry(entry)

    map =
      log_entry
      |> Timber.LogEntry.to_map!()
      |> Map.put(:message, IO.chardata_to_string(log_entry.message))

    Msgpax.pack!([map])
  end

  def add_in_memory_logger_backend(pid) when is_pid(pid) do
    {:ok, _pid} = Logger.add_backend(Timber.LoggerBackends.InMemory)
    Logger.configure_backend(Timber.LoggerBackends.InMemory, callback_pid: pid)
    :ok = Logger.remove_backend(:console)

    ExUnit.Callbacks.on_exit(fn ->
      :ok = Logger.remove_backend(Timber.LoggerBackends.InMemory)
      {:ok, _pid} = Logger.add_backend(:console)
    end)
  end
end


test/support/test_gen_server.ex
defmodule Timber.Exceptions.TestGenServer do
  use GenServer

  def start_link(pid) do
    GenServer.start_link(__MODULE__, pid)
  end

  def do_throw(pid) do
    send(pid, :throw)
  end

  def bad_exit(pid) do
    send(pid, :bad_exit)
  end

  def raise(pid) do
    send(pid, :raise)
  end

  def divide(pid, dividend, divisor) do
    send(pid, {:divide, dividend, divisor})
  end

  def divide_call(pid, dividend, divisor) do
    GenServer.call(pid, {:divide, dividend, divisor})
  end

  def init(pid) do
    {:ok, pid}
  end

  def handle_call({:divide, dividend, divisor}, _from, state) do
    {:reply, dividend / divisor, state}
  end

  def handle_info({:divide, dividend, divisor}, state) do
    (dividend / divisor)
    |> Float.to_string()

    {:ok, state}
  end

  def handle_info(:throw, _state) do
    throw("I am throwing")
  end

  def handle_info(:bad_exit, state) do
    {:stop, :bad_exit, state}
  end

  def handle_info(:raise, _state) do
    raise RuntimeError, "raised error"
  end

  def terminate(_, state) do
    send(state, :terminating)
  end
end


lib/timber_exceptions.ex
defmodule Timber.Exceptions do
  @moduledoc """
  Documentation for Timber.Exceptions.
  """

  @doc """
  Adds the Timber exception handler

  The Timber exception handler (`Timber.Exceptions.ErrorLogger`) is added as a
  report handler for `:error_logger` if it has not already been added.
  """
  @spec add_handler() :: :gen_event.add_handler_ret()
  def add_handler() do
    existing_handlers = :gen_event.which_handlers(:error_logger)

    if !(Timber.Exceptions.ErrorLogger in existing_handlers) do
      :error_logger.add_report_handler(Timber.Exceptions.ErrorLogger)
    end
  end

  @doc """
  Removes the Timber exception handler
  """
  @spec remove_handler() :: :gen_event.del_handler_ret()
  def remove_handler() do
    :error_logger.delete_report_handler(Timber.Exceptions.ErrorLogger)
  end

  @doc """
  Restores the tty output for the error logger

  The default `:error_logger_tty_h` report handler is added for `:error_logger`.
  """
  @spec restore_tty() :: :gen_event.add_handler_ret()
  def restore_tty() do
    :error_logger.tty(true)
  end

  @doc """
  Disables the default tty output for the error logger
  """
  @spec disable_tty() :: :ok
  def disable_tty() do
    :error_logger.tty(false)
  end
end


lib/timber_exceptions/translator.ex
defmodule Timber.Exceptions.Translator do
  @moduledoc """
  This module implements a Logger translator to take advantage of
  the richer metadata available from Logger in OTP 21 and Elixir 1.7+.

  Including the translator allows for crash reasons and stacktraces to be
  included as structured metadata within Timber.

  The translator depends on using Elixir's internal Logger.Translator, and
  is not compatible with other translators as a Logger event can only be
  translated once.

  To install, add the translator in your application's start function:
  ```
  # ...
  :ok = Logger.add_translator({Timber.Exceptions.Translator, :translate})

  opts = [strategy: :one_for_one, name: MyApp.Supervisor]
  Supervisor.start_link(children, opts)
  ```
  """

  @max_backtrace_size 20

  def translate(min_level, level, kind, message) do
    case Logger.Translator.translate(min_level, level, kind, message) do
      {:ok, char, metadata} ->
        new_metadata = transform_metadata(metadata)

        {:ok, char, new_metadata}

      {:ok, char} ->
        {:ok, char}

      :skip ->
        :skip

      :none ->
        :none
    end
  end

  def transform_metadata(nil), do: []

  def transform_metadata(metadata) do
    with {:ok, crash_reason} <- Keyword.fetch(metadata, :crash_reason),
         {:ok, error} <- get_error(crash_reason) do
      event = %{
        error: error
      }

      Keyword.merge([event: event], metadata)
    else
      _ ->
        metadata
    end
  end

  defp get_error({{%{__exception__: true} = error, stacktrace}, _stack})
       when is_list(stacktrace) do
    {:ok, build_error(error, stacktrace)}
  end

  defp get_error({%{__exception__: true} = error, stacktrace}) when is_list(stacktrace) do
    {:ok, build_error(error, stacktrace)}
  end

  defp get_error({{_type, reason}, stacktrace}) when is_list(stacktrace) do
    {:ok, build_error(reason, stacktrace)}
  end

  defp get_error({error, stacktrace}) when is_list(stacktrace) do
    {:ok, build_error(error, stacktrace)}
  end

  defp get_error(_) do
    {:error, :no_info}
  end

  defp build_error(%{__exception__: true, __struct__: module} = error, stacktrace) do
    message = Exception.message(error)
    module_name = Timber.Utils.Module.name(module)

    %{
      message: message,
      name: module_name,
      backtrace: build_backtrace(stacktrace)
    }
  end

  defp build_error(error, stacktrace) do
    ErlangError.normalize(error, stacktrace)
    |> build_error(stacktrace)
  end

  defp build_backtrace([trace | _] = backtrace) when is_map(trace) do
    Enum.slice(backtrace, 0..(@max_backtrace_size - 1))
  end

  defp build_backtrace([stack | _rest] = stacktrace) when is_tuple(stack) do
    stacktrace_to_backtrace(stacktrace)
  end

  defp build_backtrace(_) do
    []
  end

  defp stacktrace_to_backtrace(stacktrace) do
    # arity is an integer or list of arguments
    Enum.map(stacktrace, fn {module, function, arity, location} ->
      arity =
        case arity do
          arity when is_list(arity) -> length(arity)
          _ -> arity
        end

      file =
        Keyword.get(location, :file)
        |> Kernel.to_string()

      line = Keyword.get(location, :line)

      %{
        function: Exception.format_mfa(module, function, arity),
        file: file,
        line: line
      }
    end)
  end
end


lib/timber_exceptions/error_logger.ex
defmodule Timber.Exceptions.ErrorLogger do
  @moduledoc """
  Handles error reports from the `:error_logger` application.

  Timber can automatically log the exceptions that occur in your application as
  exception events with all the necessary metadata by registering as an
  `:error_logger` handler. To activate Timber's error logging system, you need
  to add a few configuration lines and add the `ErrorLogger` handler during
  application start:

  ```
  # Enable Timber's error capturing system
  config :timber_exceptions, :capture_errors, true

  # Disable Elixir's default error capturing system
  config :logger, :handle_otp_reports, false
  ```

  ```
  def start(_type, _opts) do
    children = [
      supervisor(MyApp.Repo, []),
      supervisor(MyAppWeb.Endpoint, [])
    ]

    opts = [strategy: :one_for_one, name: MyApp.Supervisor]

    :ok = :error_logger.add_report_handler(Timber.Exceptions.ErrorLogger)

    Supervisor.start_link(children, opts)
  end
  ```

  Users should be aware that report handlers can be added multiple times, which
  can result in duplicated logs. This concern is most relevant to umbrella
  applications or applications with multiple entry-points. To solve this, the
  handler should be added only at the primary application entry-point. If that
  is not feasible, the other solution is to check if the handler exists before
  adding it. Example:

  ```
  if !(Timber.Exceptions.ErrorLogger in :gen_event.which_handlers(:error_logger)) do
    :ok = :error_logger.add_report_handler(Timber.Exceptions.ErrorLogger)
  end
  ```

  ## Elixir Logger's OTP Report Handler

  The `Logger` application (which is distributed with Elixir) has an OTP report
  handler which logs errors and will activate it by default. However, the logs
  it writes only contain textual information without any metadata. Keeping this handler
  active will cause duplicate errors to be reported to the log, which is why
  we recommend disabling it using the following configuration option:

  ```
  config :logger, :handle_otp_reports, false
  ```

  However, the OTP report handler handles additional report types as well. If you
  find that you would like these reports to be logged, just be aware that every
  exception will be displayed twice in the log. This module also consolidates many
  error reports into a single line that by default would be logged across multiple
  lines.

  Since the OTP report handler does not add the requisite metadata, Timber's console
  will not identify the errors it logs as exception events when you search.

  ## Elixir Logger's SASL Report Handler

  The Elixir `Logger` application also comes with a SASL (System Architecture Support Libraries)
  report handler. Timber does not currently handle these reports, so activating the
  standard handler will not cause duplicate logs.

  ## :error_logger Output

  When Timber's error capturing system is activated, it will also disable `:error_logger`'s `tty`
  output. In most cases, this is what you want, otherwise, otherwise it will print out reports
  to the `tty` in a plain text (and rather ugly) format.

  If you do not want the `tty` output to be disabled, you can keep it on using the following
  config:

  ```
  config :timber, :disable_kernel_error_tty, false
  ```
  """

  @behaviour :gen_event

  alias Timber.Events.ErrorEvent

  @doc false
  def init(opts \\ []) do
    # As this is a :gen_event module, there is no built-in back-pressure mechanism.
    # To avoid holding too many messages, the process will keep up to a default of
    # 300 messages. This level is under the `:max_count` key in the state.
    # Once this level is reached, it will drop messages until it
    # is holding 75% of the maximum (225 by default). This level is under the
    # `:keep_count` key in the state. Once messages have been dropped,
    # it will process 10% of the maximum messages (the next 30 by default)
    # to avoid dropping too many consecutive messages before checking the message
    # count again. The number of messages to process during this is tracked under the
    # `:skip_count` key in the state.
    # The pid for the `Logger` process is also stored in state so messages can be
    # sent from this process to a pid instead of a named process. This is to avoid
    # crashing when sending messages to a named process.

    max_message_count = Keyword.get(opts, :max_message_count, 300)

    state = %{
      logger: Process.whereis(Logger),
      skip_count: 0,
      max_count: max_message_count,
      keep_count: trunc(max_message_count * 0.75)
    }

    {:ok, state}
  end

  @doc false
  # Ignores log events from other nodes
  def handle_event({_, gl, _}, state) when node(gl) != node() do
    {:ok, state}
  end

  def handle_event(event, state) do
    state = check_threshold(state)
    log_event(event, state)

    {:ok, state}
  end

  def get_metadata({_format, [_pid, {error, stacktrace}]}, :error) do
    case handle_error_info({error, stacktrace}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_pid, _last_message, _state, {error, stacktrace}]}, :error)
      when is_list(stacktrace) do
    case handle_error_info({error, stacktrace}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_pid, _last_message, _state, reason]}, :error) do
    case handle_error_info({reason, []}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_, _, _, {error, stacktrace}]}, :error) do
    case handle_error_info({error, stacktrace}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_, _, _, error]}, :error) do
    case handle_error_info({error, []}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_pid, _message, _pid2, {error, stack}, _pid3, _stack2]}, :error) do
    case handle_error_info({error, stack}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_pid, _pid2, _function, _args, {error, stacktrace}]}, :error) do
    case handle_error_info({error, stacktrace}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata({_format, [_module, _pid, reason]}, :warn) do
    case handle_error_info({reason, []}) do
      {:ok, event} ->
        Timber.Event.to_metadata(event)

      {:error, _} ->
        []
    end
  end

  def get_metadata(_data, _level) do
    []
  end

  defp log_event({:error, gl, {pid, format, data}}, state),
    do: do_log_event(:error, :format, gl, pid, {format, data}, state)

  defp log_event({:error_report, gl, {pid, :std_error, format}}, state),
    do: do_log_event(:error, :report, gl, pid, {:std_error, format}, state)

  defp log_event({:warning_msg, gl, {pid, format, data}}, state),
    do: do_log_event(:warn, :format, gl, pid, {format, data}, state)

  defp log_event({:warning_report, gl, {pid, :std_warning, format}}, state),
    do: do_log_event(:warn, :report, gl, pid, {:std_warning, format}, state)

  defp log_event({:info_msg, gl, {pid, format, data}}, state),
    do: do_log_event(:info, :format, gl, pid, {format, data}, state)

  defp log_event({:info_report, gl, {pid, :std_info, format}}, state),
    do: do_log_event(:info, :report, gl, pid, {:std_info, format}, state)

  defp log_event(_, _state), do: :ok

  defp do_log_event(level, kind, gl, pid, data, state) do
    %{
      mode: mode,
      level: min_level,
      utc_log: utc_log?,
      truncate: truncate,
      translators: translators
    } = Logger.Config.__data__()

    with true <- Logger.compare_levels(level, min_level) != :lt and mode != :discard,
         {:ok, message} <- translate(translators, min_level, level, kind, data, truncate) do
      meta =
        get_metadata(data, level)
        |> Keyword.put(:pid, pid)

      message = Logger.Utils.truncate(message, truncate)
      event = {Logger, message, Logger.Utils.timestamp(utc_log?), meta}
      :gen_event.notify(state.logger, {level, gl, event})
    end
  end

  def handle_info(_, state) do
    {:ok, state}
  end

  def handle_call(_, state) do
    {:ok, :ok, state}
  end

  def code_change(_old, state, _extra) do
    {:ok, state}
  end

  def terminate(_reason, _state) do
    :ok
  end

  defp handle_error_info({{%{__exception__: true} = error, stacktrace}, _stack})
       when is_list(stacktrace) do
    {:ok, build_error_event(error, stacktrace, :error)}
  end

  defp handle_error_info({%{__exception__: true} = error, stacktrace}) when is_list(stacktrace) do
    {:ok, build_error_event(error, stacktrace, :error)}
  end

  defp handle_error_info({{type, reason}, stacktrace}) when is_list(stacktrace) do
    {:ok, build_error_event(reason, stacktrace, type)}
  end

  defp handle_error_info({error, stacktrace}) when is_list(stacktrace) do
    {:ok, build_error_event(error, stacktrace, :error)}
  end

  defp handle_error_info(_) do
    {:error, :no_info}
  end

  defp build_error_event(%{__exception__: true} = error, stacktrace, _type) do
    e =
      ErrorEvent.from_exception(error)
      |> ErrorEvent.add_backtrace(stacktrace)

    # %{e | type: to_string(type)}
    e
  end

  defp build_error_event(error, stacktrace, _type) do
    e =
      ErlangError.normalize(error, stacktrace)
      |> ErrorEvent.from_exception()
      |> ErrorEvent.add_backtrace(stacktrace)

    # %{e | type: to_string(type)}
    e
  end

  defp check_threshold(%{skip_count: 0, keep_count: keep_count, max_count: max_count} = state) do
    current_count = message_count()

    if current_count >= max_count do
      drop_count = current_count - keep_count
      discard_messages(drop_count)

      message =
        "Timber.ErrorLogger has discarded #{drop_count} of #{current_count} messages in " <>
          "its inbox. The maximum number of messages is #{max_count}. " <>
          "The current number of messages is now #{keep_count}"

      %{utc_log: utc_log?} = Logger.Config.__data__()
      event = {Logger, message, Logger.Utils.timestamp(utc_log?), pid: self()}
      :gen_event.notify(state.logger, {:warn, Process.group_leader(), event})

      # Send 10% of the maximum messages before checking message count again
      %{state | skip_count: trunc(max_count * 0.1)}
    else
      state
    end
  end

  defp check_threshold(%{skip_count: skip_count} = state) do
    %{state | skip_count: skip_count - 1}
  end

  defp message_count() do
    {:message_queue_len, count} = Process.info(self(), :message_queue_len)
    count
  end

  defp discard_messages(0) do
    :ok
  end

  defp discard_messages(count) do
    receive do
      {:notify, _event} ->
        discard_messages(count - 1)
    after
      0 -> :ok
    end
  end

  # Elixir 1.6 stopped using :io_lib.format/2 in favor of :io_lib.build_text/1
  # This also included an internal replacement of Logger.Utils.inspect/3
  # with Logger.Utils.scan_inspect/3
  defmacrop build_text(format, args, truncate) do
    if(Version.match?(System.version(), "~> 1.6")) do
      quote do
        msg =
          Logger.Utils.scan_inspect(unquote(format), unquote(args), unquote(truncate))
          |> :io_lib.build_text()

        {:ok, msg}
      end
    else
      quote do
        {format, args} = Logger.Utils.inspect(unquote(format), unquote(args), unquote(truncate))
        {:ok, :io_lib.format(format, args)}
      end
    end
  end

  defp translate([{mod, fun} | t], min_level, level, kind, data, truncate) do
    case apply(mod, fun, [min_level, level, kind, data]) do
      {:ok, chardata} -> {:ok, chardata}
      :skip -> :skip
      :none -> translate(t, min_level, level, kind, data, truncate)
    end
  end

  defp translate([], _min_level, _level, :format, {format, args}, truncate) do
    build_text(format, args, truncate)
  end

  defp translate([], _min_level, _level, :report, {_type, data}, _truncate) do
    {:ok, Kernel.inspect(data)}
  end
end


mix.exs
defmodule Timber.Exceptions.MixProject do
  use Mix.Project

  @version "2.1.0"
  @source_url "https://github.com/timberio/timber-elixir-exceptions"
  @homepage_url "https://github.com/timberio/timber-elixir-exceptions"
  @project_description """
  Timber's exception tracking integration captures Elixir exceptions as
  structured data in your logs
  """

  # Package options for the Hex package listing
  #
  # See `mix help hex.publish` for more information about
  # the options used in this section
  defp package() do
    [
      name: :timber_exceptions,
      files: ["lib", "mix.exs", "README*", "LICENSE*"],
      maintainers: ["Ben Johnson", "David Antaramian"],
      licenses: ["ISC"],
      links: %{
        "GitHub" => @source_url
      }
    ]
  end

  def project do
    [
      app: :timber_exceptions,
      name: "Timber Exceptions",
      version: @version,
      elixir: "~> 1.7",
      elixirc_paths: elixirc_paths(Mix.env()),
      description: @project_description,
      source_url: @source_url,
      homepage_url: @homepage_url,
      package: package(),
      deps: deps(),
      docs: docs(),
      start_permanent: Mix.env() == :prod,
      preferred_cli_env: preferred_cli_env(),
      test_coverage: test_coverage(),
      dialyzer: dialyzer()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {Timber.Exceptions.Application, []}
    ]
  end

  # Compiler paths switched on the Mix environment
  #
  # The `lib` directory is always compiled
  #
  # In the :test environment, `test/support` will also be compiled
  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Preferred CLI Environment details
  #
  # Defines the preferred environment for Mix tasks
  defp preferred_cli_env() do
    [
      coveralls: :test,
      "coveralls.details": :test,
      "coveralls.html": :test,
      "coveralls.travis": :test
    ]
  end

  # Test Coverage configuration
  #
  # Sets the test converage tool to be Coveralls
  defp test_coverage() do
    [
      tool: ExCoveralls
    ]
  end

  # Dialyzer configuration
  defp dialyzer() do
    [
      plt_add_deps: true
    ]
  end

  # Documentation options for ExDoc
  defp docs() do
    [
      source_ref: "v#{@version}",
      main: "readme",
      logo: "doc_assets/logo.png",
      extras: [
        "README.md": [title: "README"],
        "LICENSE.md": [title: "LICENSE"]
      ]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:timber, "~> 3.1"},

      #
      # Tooling
      #

      {:credo, "~> 1.0", only: [:dev, :test]},
      {:dialyxir, "~> 0.5", only: [:dev, :test]},
      {:ex_doc, "~> 0.19.0", only: [:dev]},
      {:excoveralls, "~> 0.10", only: [:dev, :test]}
    ]
  end
end
